### PR TITLE
docs: Add remaining Keyman Engine for Android API's for KMManager

### PR DESCRIPTION
**Pull request by darcywong00**
This adds remaining API pages to KMManager pointed out in an earlier review comment

https://github.com/keymanapp/help.keyman.com/blob/73c9bc1991f56b9dc99fbe2eea924debf40be25d/developer/engine/android/18.0/KMManager/index.md?plain=1#L179-L180

              Is this missing a separate page?

_Originally posted by @mcdurdin in https://github.com/keymanapp/help.keyman.com/pull/1268#discussion_r1604385682_

### CODE REVIEW

File: developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+| Ctrl         | `0x20`   | Control Flag                           |
+| Alt          | `0x40`   | Alt Flag                               |
+
+`keyboard`
+: `KeyboardType.KEYBOARD_TYPE_INAPP` or `KeyboardType.KEYBOARD_TYPE_SYSTEM`
+
+`lstates`
+: lock state
+
+| value  | Description    |
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md:37]:
```suggestion
| Value  | Description    |
```

File: developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+|--------|----------------|
+| 0x0200 | No caps        |
+| 0x0400 | Num lock       |
+| 0x0800 | No Num         |
+| 0x1000 | Scroll         |
+| 0x2000 | No scroll lock |
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md:43]: Caps lock was missing!

```suggestion
| `0x0100` | Caps lock      |
| `0x0200` | No caps lock   |
| `0x0400` | Num lock       |
| `0x0800` | No Num lock    |
| `0x1000` | Scroll         |
| `0x2000` | No scroll lock |
```

File: developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+| Right Alt    | `0x08`   | Right Alt Flag                         |
+| Shift        | `0x10`   | Shift Flag                             |
+| Ctrl         | `0x20`   | Control Flag                           |
+| Alt          | `0x40`   | Alt Flag                               |
+
+`keyboard`
+: `KeyboardType.KEYBOARD_TYPE_INAPP` or `KeyboardType.KEYBOARD_TYPE_SYSTEM`
+
+`lstates`
+: lock state
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md:35]:
```suggestion
: lock state. If neither the on or off state is specified for a lock key, then it is treated as unknown state for keyboard rule matching.
```

File: developer/engine/android/18.0/KMManager/getKMKeyboard.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+```
+### Parameters 
+type
+: `KeyboardType.KEYBOARD_TYPE_INAPP` or `KeyboardType.KEYBOARD_TYPE_SYSTEM`
+
+### Returns
+Returns the KMKeyboard object. 
+
+## Description
+The `getKMKeyboard()` method returns the KMKeyboard for the specified keyboard type.
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getKMKeyboard.md:20]: KMKeyboard should link to the type

File: developer/engine/android/18.0/KMManager/getLexicalModelInfo.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+### Parameters
+
+`context`
+: The context.
+
+`index`
+: Non-negative position of the installed lexical model
+
+### Returns
+Returns a dictionary containing information on the specified lexical model.
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getLexicalModelInfo.md:24]: The dictionary should be documented here (and the type `Hashmap<String, String>`

File: developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+## Examples
+
+### Example: Using `executeHardwareKeystroke()`
+The following script illustrate the use of `executeHardwareKeystroke()`:
+
+```java
+    // Send keystroke to KeymanWeb for processing: will return true to swallow the keystroke
+    return KMManager.executeHardwareKeystroke(code, keymanModifiers, keyboardType, Lstates, androidModifiers);
+```
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md:62]: This is not a very helpful example, though -- where do we collect the various values from? Can we use an example which ties into an Android event?

File: developer/engine/android/18.0/KMManager/getLexicalModelInfo.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+KMManager.getLexicalModelInfo(Context context, int index)
+```
+
+### Parameters
+
+`context`
+: The context.
+
+`index`
+: Non-negative position of the installed lexical model
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getLexicalModelInfo.md:21]: Zero-based index?

File: developer/engine/android/18.0/KMManager/getLexicalModelInfo.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+### Returns
+Returns a dictionary containing information on the specified lexical model.
+
+
+## Description
+Use this method to get details about a sepcified lexical mode. (language and keyboard information)
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getLexicalModelInfo.md:28]:
```suggestion
Use this method to get details about a specified lexical model. (language and keyboard information)
```

File: developer/engine/android/18.0/KMManager/getLexicalModelInfo.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+---
+
+## Summary
+
+The `getLexicalModelInfo()` method returns returns dictionary information of the specified lexical model.
+
+## Syntax
+
+```java
+KMManager.getLexicalModelInfo(Context context, int index)
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getLexicalModelInfo.md:12]:
```suggestion
HashMap<String,String> KMManager.getLexicalModelInfo(Context context, int index)
```

File: developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
+---
+title: KMManager.executeHardwareKeystroke()
+---
+
+## Summary
+The **executeHardwareKeystroke()** method passes the keystroke generated from a physical keyboard to the KeymanWeb engine to handle.
+
+## Syntax
+```java
+KMManager.executeHardwareKeystroke(int code, int shift, KeyboardType keyboard, int lstates, int eventModifiers)
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md:10]:
```suggestion
bool KMManager.executeHardwareKeystroke(int code, int shift, KeyboardType keyboard, int lstates, int eventModifiers)
```

File: developer/engine/android/18.0/KMManager/getKMKeyboard.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
+---
+title: KMManager.getKMKeyboard()
+---
+
+## Summary
+The `getKMKeyboard()` method returns the KMKeyboard type depending whether it's an in-app or system keyboard.
+
+## Syntax
+```java
+KMManager.getKMKeyboard(KeyboardType type)
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getKMKeyboard.md:10]:
```suggestion
KMKeyboard KMManager.getKMKeyboard(KeyboardType type)
```

File: developer/engine/android/18.0/KMManager/getKMKeyboard.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+```
+### Parameters 
+type
+: `KeyboardType.KEYBOARD_TYPE_INAPP` or `KeyboardType.KEYBOARD_TYPE_SYSTEM`
+
+### Returns
+Returns the KMKeyboard object.
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getKMKeyboard.md:17]: Link to the KMKeyboard type?

File: developer/engine/android/18.0/KMManager/getLexicalModelsList.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
+---
+title: KMManager.getLexicalModelsList()
+---
+
+## Summary
+The **getLexicalModelsList()** method returns the array of lexical models list.
+
+## Syntax
+```java
+KMManager.getLexicalModelsList(Context context)
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getLexicalModelsList.md:10]:
```suggestion
ArrayList<HashMap<String, String>> KMManager.getLexicalModelsList(Context context)
```

File: developer/engine/android/18.0/KMManager/getLexicalModelsList.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+### Parameters
+`context`
+: The context.
+
+### Returns
+Returns a list of all the installed lexical models.
+
+## Description
+Use this method to get details of all the installed lexical models.
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/getLexicalModelsList.md:21]: The dictionary should be documented, perhaps link to `getLexicalModelInfo`?

File: developer/engine/android/18.0/KMManager/lexicalModelExists.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
+---
+title: KMManager.lexicalModelExists()
+---
+
+## Summary
+The **lexicalModelExists()** method returns whether the specified lexical model exists in lexical models list.
+
+## Syntax
+```java
+KMManager.lexicalModelExists(Context context, String packageID, String languageID, String modelID)
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/lexicalModelExists.md:10]:
```suggestion
bool KMManager.lexicalModelExists(Context context, String packageID, String languageID, String modelID) 
```

File: developer/engine/android/18.0/KMManager/lexicalModelExists.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+## Description
+Use this method to determine if a lexical model with given package ID, language ID, and model ID exists in the lexical models list.
+
+## Examples
+
+### Example: Using lexicalModelExists()
+The following script illustrate the use of `lexicalModelExists()`:
+```java
+    // Determine if the specified lexical model exists for English (en)
+    boolean exists = KMManager.lexicalModelExists(context, "nrc", "en", "en.mtnt";
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/lexicalModelExists.md:38]:
```suggestion
    boolean exists = KMManager.lexicalModelExists(context, "nrc.en.mtnt", "en", "nrc.en.mtnt");
```

I am not 100% sure of this, but I don't think the original example was right?

File: developer/engine/android/18.0/KMManager/lexicalModelExists.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+## See also
+* [getLexicalModelInfo](getLexicalModelInfo())
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/lexicalModelExists.md:42]:
```suggestion
* [`getLexicalModelInfo()`](getLexicalModelInfo)
```

File: developer/engine/android/18.0/KMManager/setCanRemoveKeyboard.md [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+---
+
+## Summary
+The **setCanRemoveKeyboard()** method sets whether removing a keyboard is allowed, like in the keyboard picker menu.
+
+## Syntax
+
+```javascript
+    KMManager.setCanRemoveKeyboard(boolean newValue)
+```
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/setCanRemoveKeyboard.md:12]:
````suggestion
```java
    void KMManager.setCanRemoveKeyboard(boolean newValue)
```
````

File: developer/engine/android/18.0/KMManager/setCanRemoveKeyboard.md [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+## Parameters
+
+`newValue`
+: `true` if removing a keyboard is allowed, `false` otherwise.
+
+## Description
+Use this method to grant Keyman Engine for Android permission to remove keyboards from the keyboard list. If set to `false`, the installed keyboard list is permanent.
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/setCanRemoveKeyboard.md:20]: Is this to allow the end user to remove the keyboard? I'm not clear

File: developer/engine/android/18.0/KMManager/setCanRemoveKeyboard.md [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+### Example: Using `setCanRemoveKeyboard()`
+
+The following script illustrates the use of `setCanRemoveKeyboard()`:
+```java
+    // Disable removing keyboards
+    KMManager.setCanRemoveKeyboard(false);
+```
+
+## See also
+* [canRemoveKeyboard()](canRemoveKeyboard)
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMManager/setCanRemoveKeyboard.md:33]:
```suggestion
* [`canRemoveKeyboard()`](canRemoveKeyboard)
```

**Review by mcdurdin — CHANGES_REQUESTED** [review]
A fair few typos; so sorry for all the comments!

File: developer/engine/android/18.0/KMManager/getLexicalModelInfo.md (hunk 7 above)
Comment (darcywong00) [reply developer/engine/android/18.0/KMManager/getLexicalModelInfo.md:21]:
```suggestion
`index`
: Zero-based index of the installed lexical model
```

File: developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md (hunk 6 above)
Comment (darcywong00) [reply developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md:63]: These look to be determined from `(int keyCode, KeyEvent event)`

https://github.com/keymanapp/keyman/blob/e4b0d1859c5517db0dd963a283a364ca2f284425/android/KMEA/app/src/main/java/com/keyman/engine/KMHardwareKeyboardInterpreter.java#L24-L74

File: developer/engine/android/18.0/KMManager/getKMKeyboard.md (hunk 12 above)
Comment (darcywong00) [reply developer/engine/android/18.0/KMManager/getKMKeyboard.md:17]: Hm, that might be a separate PR. I'm not sure I saw a page for KMKeyboard.
(To put at /developer/engine/android/18.0/#toc-classes ?)

File: developer/engine/android/18.0/KMManager/getKMKeyboard.md (hunk 4 above)
Comment (darcywong00) [reply developer/engine/android/18.0/KMManager/getKMKeyboard.md:20]: Added KMKeyboard page and link in 7f99859dd47d3bc10e728dfd695a5599b3d9f20c

**Comment by darcywong00** [comment]
I think I've addressed all the comments so far. Ready to re-review

File: developer/engine/android/18.0/KMKeyboard/index.md [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+```java
+KMKeyboard.methodName()
+```
+
+## Description
+
+The KMKeyboard is the core class which provides most of the methods you will need to develop an in-app or system keyboard with Keyman Engine
+
+<!-- Fill in examples of how to use, etc. -->
```
Comment (mcdurdin) [inline developer/engine/android/18.0/KMKeyboard/index.md:19]: TODO

File: developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md (hunk 6 above)
Comment (mcdurdin) [reply developer/engine/android/18.0/KMManager/executeHardwareKeystroke.md:63]: Hmm, should this be a TODO for future improvement then?

File: developer/engine/android/18.0/KMManager/getKMKeyboard.md (hunk 12 above)
Comment (mcdurdin) [reply developer/engine/android/18.0/KMManager/getKMKeyboard.md:17]: TODO for future improvement

File: developer/engine/android/18.0/KMManager/getLexicalModelsList.md (hunk 14 above)
Comment (mcdurdin) [reply developer/engine/android/18.0/KMManager/getLexicalModelsList.md:21]: This is still TODO

**Review by mcdurdin — APPROVED** [review]
Approving but I think there are future TODOs. Ideally, the examples will include surrounding code to help clarify when and where a function may be called

**Pull request merged**